internal/schemahelpers/schemahelpers.go

```diff
@@ -0,0 +1,167 @@
+package schemahelpers
+
+import (
+	"log"
+
+	datasource_schema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	resource_schema "github.com/hashicorp/terraform-plugin-framework/resource/schema"
+)
+
+func DataSourceSchemaFromResourceSchema(resourceSchema map[string]resource_schema.Attribute, idField string) map[string]datasource_schema.Attribute {
```

Comment:
Regarding the schema duplication issue, I tried to just duplicate things at first but turns out there is a lot of schema to be duplicated. Particularly, all the destination authentication methods mechanism, each requiring its own schema that must be duplicated.

After some further exploration, I was able to write a conversion function where it takes the resource schema and converts to data source schema. I found a snippet online from another [provider](https://github.com/Doridian/terraform-provider-hexonet/blob/main/hexonet/utils/schema_helpers.go) (MIT license of course) and expanded upon it to make it works for our use case.

From there, everything is pretty straightforward. I refactored each service's `resource.go`, `datasource.go`, and `schema.go` files to make things cleaner and follow a particular template. Hope everything makes sense and please let me know if you have any input here!